File: src/robustCorr.c
/* 
   Function to calculate robust correlation coeficients (minimizing the
   influences of one outlier). This strategy uses an idea similar to
   leave-one-out of cross-validation, where we make a decision based on the
   "rule of three numbers" thought by Diana an me.
   
   Gustavo H. Esteves
   22/05/07 
*/


#include <R.h>
#include "stats.h"


void corR(double *x, double *y, int n, double *res) {
    
    int i, j, k, *idx;
    double *tmp, *tmp1, *tmp2, sd1, sd2, median;
    
    idx = (int *) R_alloc(n, sizeof(int));
    tmp = (double *) R_alloc(n, sizeof(double));
    tmp1 = (double *) R_alloc((n-1), sizeof(double));
    tmp2 = (double *) R_alloc((n-1), sizeof(double));
    
    for(j=0; j<n; j++) {
	k = 0;
	for(i=0; i<n; i++)
	    if(i != j) {
		tmp1[k] = x[i];
		tmp2[k] = y[i];
		k += 1;
	    }
	sd1 = stats_sd(tmp1, (n-1));
	sd2 = stats_sd(tmp2, (n-1));
	tmp[j] = stats_covariance(tmp1, tmp2, (n-1))/(sd1*sd2);

	idx[j] = j;

    }
    
    /* Making the decision based on the rule of three numbers */

    rsort_with_index(tmp, idx, n);
    
    median = stats_median_from_sorted_data(tmp, n);
    
    if((tmp[n-1]-median) < (median-tmp[0])) {
	res[0] = tmp[0];
	res[1] = (double)idx[0];
    }
    else {
	res[0] = tmp[n-1];
	res[1] = (double)idx[n-1];
    }
    
}


void robustCorr(double *x, int *ng, int *ns, double *c, int *idx) {
  
    int i, j, k;
    double tmp[2], *tmp1, *tmp2;
    
    tmp1 = (double *) R_alloc((*ns), sizeof(double));
    tmp2 = (double *) R_alloc((*ns), sizeof(double));
    
    for (i=0; i<(*ng-1); i++)
	for(j=(i+1); j<*ng; j++) {
	    for(k=0; k<*ns; k++) {
		tmp1[k] = x[((*ns)*i)+k];
		tmp2[k] = x[((*ns)*j)+k];
	    }
	    corR(tmp1, tmp2, *ns, tmp);
	    c[((*ng)*i)+j] = tmp[0];
	    c[((*ng)*j)+i] = tmp[0];
	    idx[((*ng)*i)+j] = (int)tmp[1]+1;
	    idx[((*ng)*j)+i] = (int)tmp[1]+1;
	}
    
}

File: R/robustCorr.R
## Function to calculate robust correlation coeficients (minimizing the
## influences of one outlier). This strategy uses an idea similar to
## leave-one-out of cross-validation, where we make a decision based on the
## "rule of three numbers".
##
## Parameters: x -> matrix to calculate the correlations between all
##                  pairs of rows from x. Return a square matrix with
##                  all correlations. If x is a matrix, the function
##                  calculates a matrix of robust correlations, else
##                  x must be a vector and y must be specified as
##                  another vector of same length and the correlation
##                  between both are calculate
##             y -> optional numeric vector
##
## Gustavo H. Esteves
## 15/05/07
##
## Version: 1.0
##


robustCorr <- function(x, y=NULL) {
    
    
    if(length(dim(x)) == 2) {
        
        ## Getting the dimensions of the matrix
        ns <- dim(x)
        
        ## Doing the cor calculations from C compiled code
        tmp <- .C("robustCorr", as.double(as.vector(t(x))), as.integer(ns[1]),
        as.integer(ns[2]), as.double(rep(1, (ns[1])^2)), as.integer(rep(0,
        (ns[1])^2)), DUP=FALSE, PACKAGE="maigesPack")
        
        
        ## Defining the result matrix
        tmp1 <- matrix(tmp[[4]], ns[1], ns[1], TRUE)
        tmp2 <- matrix(tmp[[5]], ns[1], ns[1], TRUE)
        if(!is.null(rownames(x))) {
            rownames(tmp1) <- rownames(x)
            colnames(tmp1) <- rownames(x)
            rownames(tmp2) <- rownames(x)
            colnames(tmp2) <- rownames(x)
        }
        
        return(list(Rcor=tmp1, idx=tmp2))
        
    }
    else if(is.vector(x) & !is.null(y)) {
        
        ## Doing simple tests
        if(!is.vector(y))
            stop("y must be a vector of same length as x!")
        if(length(x) != length(y))
            stop("x and y must have same length!")
        
        ## calculating the correlation value
        calc <- rbind(x, y)
        ns <- dim(calc)
        
        tmp <- .C("robustCorr", as.double(as.vector(t(calc))),
        as.integer(ns[1]), as.integer(ns[2]), as.double(rep(1, (ns[1])^2)),
        as.integer(rep(0, (ns[1])^2)), DUP=FALSE, PACKAGE="maigesPack")
        
        
        ## Defining the result
        tmp1 <- matrix(tmp[[4]], ns[1], ns[1], TRUE)
        tmp2 <- matrix(tmp[[5]], ns[1], ns[1], TRUE)
        
        return(list(Rcor=tmp1[1,2], idx=tmp2[1,2]))
        
    }
    else
        stop("x must be a numerical matrix or vector, case where you must
        specify another vector y of same length.")
    
}
